scanner.py

```diff
@@ -1714,23 +1719,74 @@ def _is_faulty_coordinate(self, x, y, add_offsets=False):
         # Streaming mode
 
     def _check_hardware(self, sample):
+        # Validate sample input
+        if "data" not in sample or "freq" not in sample:
+            raise ValueError("Sample must contain 'data' and 'freq' keys.")
```

Comment:
I think it's better to use a klipper error for these cases.
I think the `error` defined in `mcu` is appropriate.
https://github.com/Klipper3d/klipper/blob/80d185c94c8dbec7f19d791995242af1bb197057/klippy/mcu.py#L9

Reply:
see 
4fc5ede